### PR TITLE
Fogl 5468:  new C++ purge process to handle STATISTICS_HISTORY, AUDIT TRAILS and TASKS tables

### PR DESCRIPTION
**Pull request by sstefano69**
Service Name: purge_system

Tables
- fledge.statistics_history
- fledge.tasks 
- fledge.log 

SQLite
```
year|day|key|value
2020|2020-04-01|KEY_03|3
2021|2021-05-01|KEY_01|3
2021|2021-06-01|KEY_02|10
2021|2021-06-02|KEY_01|5
```
Postgress
```
 year |          day           |  key   | value 
------+------------------------+--------+-------
 2020 | 2020-04-01 00:00:00+00 | KEY_03 |     3
 2021 | 2021-05-01 00:00:00+00 | KEY_01 |     3
 2021 | 2021-06-01 00:00:00+00 | KEY_02 |    10
 2021 | 2021-06-02 00:00:00+00 | KEY_01 |     5
(4 rows)
```
Test case
```

--- statistics_history
INSERT INTO statistics_history (id, key, history_ts, value, ts) VALUES (2000, 'KEY_01', '2021-06-10 14:49:34.392', 1, '2021-06-14 14:49:34.392');

INSERT INTO statistics_history (id, key, history_ts, value, ts) VALUES (2001, 'KEY_01', '2021-06-02 01:49:34.392', 2, '2021-06-14 14:49:34.392');
INSERT INTO statistics_history (id, key, history_ts, value, ts) VALUES (2002, 'KEY_01', '2021-06-02 02:49:34.392', 3, '2021-06-14 14:49:34.392');

INSERT INTO statistics_history (id, key, history_ts, value, ts) VALUES (2003, 'KEY_02',   '2021-06-01 14:49:34.392',10, '2021-06-14 14:49:34.392');

INSERT INTO statistics_history (id, key, history_ts, value, ts) VALUES (2004, 'KEY_01', '2021-05-01 14:49:34.392', 3, '2021-06-14 14:49:34.392');

INSERT INTO statistics_history (id, key, history_ts, value, ts) VALUES (2005, 'KEY_03', '2020-04-01 14:49:34.392', 3, '2021-06-14 14:49:34.392');

 
---- statistics_history_daily   -------------------------------
year|day|key|value
2020|2020-04-01|KEY_03|3
2021|2021-05-01|KEY_01|3
2021|2021-06-01|KEY_02|10
2021|2021-06-02|KEY_01|5


 

```

current usage
```
STATISTICS_HISTORY................................ 84          41.8%
TASKS............................................. 22          10.9%
LOG............................................... 7            3.5%

```

Delete operations  - using seconds for debug
```
query :DELETE FROM fledge.statistics_history WHERE history_ts < datetime('now', '-7 days');:
query :DELETE FROM fledge.tasks WHERE start_time < datetime('now', '-30 days');:
query :DELETE FROM fledge.log WHERE ts < datetime('now', '-30 days');:
```

logs:
```
purge_system[18976]: INFO: xxx2 PurgeSystem running
purge_system[18976]: DEBUG: xxx2 historicizeData - historicizing :statistics_history: retention days :7:
purge_system[18976]: DEBUG: xxx2 purgeTable - purging :statistics_history: retention days :7: conditionValue :7:
purge_system[18976]: DEBUG: xxx2 purgeTable - statistics_history rows purged :462:
purge_system[18976]: DEBUG: xxx2 purgeTable - purging :tasks: retention days :30: conditionValue :30:
purge_system[18976]: DEBUG: xxx2 purgeTable - tasks rows purged :76:
purge_system[18976]: DEBUG: xxx2 purgeTable - purging :log: retention days :30: conditionValue :30:
purge_system[18976]: DEBUG: xxx2 purgeTable - log rows purged :14:
purge_system[18976]: INFO: xxx2 PurgeSystem completed
```

Configuration
```
curl -sX GET http://localhost:8081/$dia_system/category/purge_system  | jq "."

{
  "retainStatsHistory": {
    "description": "This is the measure of how long to retain statistics history data for and should be measured in days.",
    "type": "integer",
    "order": "1",
    "displayName": "Retain Stats History Data (In Days)",
    "minimum": "1",
    "default": "7",
    "value": "7"
  },
  "retainAuditLog": {
    "description": "This is the measure of how long to retain audit trail information for and should be measured in days.",
    "type": "integer",
    "order": "2",
    "displayName": "Retain Audit Trail Data (In Days)",
    "minimum": "1",
    "default": "30",
    "value": "30"
  },
  "retainTaskHistory": {
    "description": "This is the measure of how long to retain task history information for and should be measured in days.",
    "type": "integer",
    "order": "3",
    "displayName": "Retain Task History Data (In Days)",
    "minimum": "1",
    "default": "30",
    "value": "30"
  }
}
--- 
```

historicization - actual query of the storage engine
```
SELECT 
sum("value") AS "sum_value", 
date(history_ts), 
key 
FROM fledge.statistics_history 
WHERE history_ts < datetime('now', '-7 seconds', 'localtime') 
GROUP BY date(history_ts), key;
```

```
---- statistics_history_daily   -------------------------------
year|day|key|value
2021|2021-06-12|BUFFERED|0
2021|2021-06-12|DISCARDED|0
2021|2021-06-12|North_Readings_to_PI|19000
2021|2021-06-12|PURGED|0
2021|2021-06-12|READINGS|69727
2021|2021-06-12|Readings Sent|19000
2021|2021-06-12|SIN_5468_ASSET_1|69727
2021|2021-06-12|UNSENT|0
2021|2021-06-12|UNSNPURGED|0

2021|2021-06-14|BUFFERED|0
2021|2021-06-14|DISCARDED|0
2021|2021-06-14|North_Readings_to_PI|278060
2021|2021-06-14|PURGED|0
2021|2021-06-14|READINGS|225055
2021|2021-06-14|Readings Sent|278060
2021|2021-06-14|SIN_5468_ASSET_1|225055
2021|2021-06-14|UNSENT|0
2021|2021-06-14|UNSNPURGED|0

```

image

schedule
```
VALUES ( 'd37265f0-c83a-11eb-b8bc-0242ac130003', -- id
         'purge_system',                         -- schedule_name
         'purge_system',                         -- process_name
         3,                                      -- schedule_type (interval)
         NULL,                                   -- schedule_time
         '23:59:00',                             -- schedule_interval (evey 24 hours)
         't',                                    -- exclusive
         'f'                                     -- enabled
       );
```

Error samples
```
localhost,Fledge,--- purge_system[8928]: INFO: xxx2 PurgeSystem starting - parameters name :purge_system:
localhost,Fledge,--- purge_system[8928]: DEBUG: xxx2 configurationHandling - categoryName :purge_system:
localhost,Fledge,--- purge_system[8928]: INFO: xxx2 configuration retainStatsHistory :7: retainAuditLog :30: retainTaskHistory :30:
localhost,Fledge,--- purge_system[8928]: INFO: xxx2 PurgeSystem running
localhost,Fledge,--- purge_system[8928]: INFO: xxx2 PurgeSystem completed
```
```
localhost,Fledge,--- purge_system[7840]: ERROR: An error occurred during the execution :Unable to find storage service at 127.0.0.1:42723:

```

### CODE REVIEW

File: C/tasks/purge_system/purge_system.cpp [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+using namespace std;
+
+volatile std::sig_atomic_t signalReceived = 0;
+
+static const string DEFAULT_CONFIG = QUOTE({
+	"retainStatsHistory": {
+		"description": "This is the measure of how long to retain statistics history data for and should be measured in days.",
+		"type": "integer",
+		"default": "7",
+		"displayName": "Retain Stats History Data (In Days)",
```
Comment (MarkRiddoch) [inline C/tasks/purge_system/purge_system.cpp:28]: This needs to be shorter to fit the GUI really, the GUI does not cope well with long strings in display names. I would suggests "Statistics Retention"

File: C/tasks/purge_system/purge_system.cpp [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+#include <thread>
+#include <csignal>
+
+using namespace std;
+
+volatile std::sig_atomic_t signalReceived = 0;
+
+static const string DEFAULT_CONFIG = QUOTE({
+	"retainStatsHistory": {
+		"description": "This is the measure of how long to retain statistics history data for and should be measured in days.",
```
Comment (MarkRiddoch) [inline C/tasks/purge_system/purge_system.cpp:25]: I would make this something like "The number of days for which full granularity statistics history is maintained."

File: C/tasks/purge_system/purge_system.cpp [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+		"type": "integer",
+		"default": "7",
+		"displayName": "Retain Stats History Data (In Days)",
+		"order": "1",
+		"minimum": "1"
+	},
+	"retainAuditLog" : {
+		"description": "This is the measure of how long to retain audit trail information for and should be measured in days.",
```
Comment (MarkRiddoch) [inline C/tasks/purge_system/purge_system.cpp:33]: "The number of days for which audit trail data is retained"

File: C/tasks/purge_system/purge_system.cpp [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+		"type": "integer",
+		"default": "30",
+		"displayName": "Retain Audit Trail Data (In Days)",
```
Comment (MarkRiddoch) [inline C/tasks/purge_system/purge_system.cpp:36]: "Audit Retention"

File: C/tasks/purge_system/purge_system.cpp [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+		"order": "2",
+		"minimum": "1"
+	},
+	"retainTaskHistory" : {
+		"description": "This is the measure of how long to retain task history information for and should be measured in days.",
```
Comment (MarkRiddoch) [inline C/tasks/purge_system/purge_system.cpp:41]: "The number of days for which task history is retained"

File: C/tasks/purge_system/purge_system.cpp [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+		"type": "integer",
+		"default": "30",
+		"displayName": "Retain Task History Data (In Days)",
```
Comment (MarkRiddoch) [inline C/tasks/purge_system/purge_system.cpp:44]: "Task Retention"

**Review by MarkRiddoch — COMMENTED** [review]
Just a few things around the configuration category, otherwise looks good.

File: C/tasks/purge_system/purge_system.cpp (hunk 1 above)
Comment (sstefano69) [reply C/tasks/purge_system/purge_system.cpp:28]: done

File: C/tasks/purge_system/purge_system.cpp (hunk 2 above)
Comment (sstefano69) [reply C/tasks/purge_system/purge_system.cpp:25]: done

File: C/tasks/purge_system/purge_system.cpp (hunk 3 above)
Comment (sstefano69) [reply C/tasks/purge_system/purge_system.cpp:33]: done

File: C/tasks/purge_system/purge_system.cpp (hunk 4 above)
Comment (sstefano69) [reply C/tasks/purge_system/purge_system.cpp:36]: done

File: C/tasks/purge_system/purge_system.cpp (hunk 5 above)
Comment (sstefano69) [reply C/tasks/purge_system/purge_system.cpp:41]: done

File: C/tasks/purge_system/purge_system.cpp (hunk 6 above)
Comment (sstefano69) [reply C/tasks/purge_system/purge_system.cpp:44]: done

**Review by MarkRiddoch — APPROVED** [review]

**Pull request merged**